### PR TITLE
Spherical Harmonics

### PR DESCRIPTION
**Pull request by markj997**
## Description

Implementation of necessary functions, constraints, examples, etc, enabling use of spherical harmonics for optimisation and solving of equilibria.

### CODE REVIEW

File: bluemira/equilibria/flux_surfaces.py [hunk 1]
```diff
@@ -871,3 +873,102 @@ def poloidal_angle(Bp_strike, Bt_strike, gamma):
...
+            * const.mu_0
+            * (r_t / r_f) ** degree
+            * np.sin(theta_f)
+            * lpmv(1, degree, np.cos(theta_f))
+            / np.sqrt(degree * (degree + 1))
+        )
+
+    amplitudes = np.matmul(currents2harmonics, I_f)
+
+    return amplitudes
```
Comment (je-cook) [inline bluemira/equilibria/flux_surfaces.py:919]: Just fyi the '@' symbol I think accomplished the same thing (uses matmul under the hood) also you could just directly return the calculation eg 
```suggestion
    return currents2harmonics @ I_f
```

File: bluemira/equilibria/opt_constraints.py [hunk 2]
```diff
@@ -61,6 +62,44 @@ def _get_dummy_equilibrium(equilibrium):
...
+    max_degree: float
+        Maximum degree of spherical harmonics desired to constrain.
+    """
+    # Does the eq get updated at each iteration, requiring I_f as input or can
+    # we just extract I_f from the eq object within the function?
+
+    # Could look at a way of adding harmonics until a desired precision in core
+    # plasma is reached.
+    for coil in eq.coilset:
+        x_f, z_f = coil.x, coil.z
```
Comment (je-cook) [inline bluemira/equilibria/opt_constraints.py:89]: Is this not just getting the x and z of the last coil in the coilset?
```python
coil = eq.coilset[-1]
x_f, z_f = coil.x, coil.z
```

**Review by je-cook — CHANGES_REQUESTED** [review]
Looks good! Just some code styling things mostly, also one about where the loop should start

File: bluemira/equilibria/opt_constraints.py [hunk 3]
```diff
@@ -61,6 +62,44 @@ def _get_dummy_equilibrium(equilibrium):
...
+        Equilibrium used to for coilset.
+    r_t: float
+        Typical length scale of the problem (e.g. radius at outer midplane)
+    initial_harmonics: np.ndarry
+        Initial harmonic amplitudes obtained from desired core plasma
+    max_degree: float
+        Maximum degree of spherical harmonics desired to constrain.
+    """
+    # Does the eq get updated at each iteration, requiring I_f as input or can
+    # we just extract I_f from the eq object within the function?
```
Comment (sebkahn) [inline bluemira/equilibria/opt_constraints.py:84]: The equilibrium gets update at each grad-Shafranov iteration by definition.

As far as I understand the GS does not get updated with the PF optimization, however the coilset by definition should. Although I am not fully aware of the implentation details @alexanderianblair ?

**Review by sebkahn — CHANGES_REQUESTED** [review]

File: bluemira/equilibria/flux_surfaces.py [hunk 4]
```diff
@@ -32,7 +32,9 @@
 import matplotlib.pyplot as plt
 import numba as nb
 import numpy as np
+import scipy.constants as const
```
[the remark this answers is not in the document]
Comment (je-cook) [reply bluemira/equilibria/flux_surfaces.py:35]: bluemira ones please. They all come from the same place which is the 2018 codata database. We get that via Pint just so we get them all from the same package

File: bluemira/equilibria/opt_constraints.py (hunk 2 above)
Comment (je-cook) [reply bluemira/equilibria/opt_constraints.py:89]: Cant remember, but I can make it so in the refactor. For now your suggestion is probably the right way forward

File: bluemira/equilibria/flux_surfaces.py [hunk 5]
```diff
@@ -871,3 +873,102 @@ def poloidal_angle(Bp_strike, Bt_strike, gamma):
     sin_theta = num / Bp_strike
 
     return np.rad2deg(np.arcsin(sin_theta))
+
+
+def harmonic_amplitude(x_f, z_f, I_f, max_degree, r_t, cylindrical=False):
+    """
+
```
[the remark this answers is not in the document]
Comment (je-cook) [reply bluemira/equilibria/flux_surfaces.py:880]: https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-math :slightly_smiling_face:

**Comment by sebkahn** [comment]
Good first pass, we should put the constraints on `opt_constraitns_funcs.py`

**Comment by CoronelBuendia** [comment]
Closing in favour of #1963

**Pull request closed without merging**